tests/io/test_partitioned_dataset.py

```diff
@@ -286,17 +291,56 @@ def test_invalid_dataset_config(self, dataset_config, error_pattern):
     @pytest.mark.parametrize(
         "dataset_config",
         [
-            {"type": CSVDataSet, "versioned": True},
-            {"type": "pandas.CSVDataSet", "versioned": True},
+            {**ds_config, "versioned": True}
+            for ds_config in LOCAL_DATASET_DEFINITION
+            if isinstance(ds_config, dict)
         ],
     )
-    def test_versioned_dataset_not_allowed(self, dataset_config):
-        pattern = (
-            "`PartitionedDataSet` does not support versioning of the underlying "
-            "dataset. Please remove `versioned` flag from the dataset definition."
+    @pytest.mark.parametrize(
+        "suffix,expected_num_parts", [("", 5), (".csv", 3), ("p4", 1)]
+    )
+    def test_versioned_dataset_save_and_load(
+        self,
+        filepath_csvs,
+        dataset_config,
+        suffix,
+        expected_num_parts,
+        partitioned_data_pandas,
+    ):
+        """Test that saved and reloaded data matches the original one for
+        the versioned data set."""
+        PartitionedDataSet(filepath_csvs, dataset_config).save(partitioned_data_pandas)
+
+        pds = PartitionedDataSet(filepath_csvs, dataset_config, filename_suffix=suffix)
+        loaded_partitions = pds.load()
+
+        assert len(loaded_partitions.keys()) == expected_num_parts
+        for partition_id, load_func in loaded_partitions.items():
+            df = load_func()
+            assert_frame_equal(df, partitioned_data_pandas[partition_id + suffix])
+            if suffix:
+                assert not partition_id.endswith(suffix)
+
+    def test_malformed_versioned_path(self, tmp_path):
+        local_dir = tmp_path / "files"
+        local_dir.mkdir()
+
+        path = local_dir / "path/to/folder/new/partition/version/partition/file"
+        path.parent.mkdir(parents=True, exist_ok=True)
+        path.write_text("content")
+
+        pds = PartitionedDataSet(
+            str(local_dir / "path/to/folder"),
+            {"type": "pandas.CSVDataSet", "versioned": True},
         )
-        with pytest.raises(DataSetError, match=re.escape(pattern)):
-            PartitionedDataSet(str(Path.cwd()), dataset_config)
+
+        pattern = re.escape(
+            f"`{path.as_posix()}` is not a well-formed versioned path ending with "
+            "`filename/timestamp/filename` (got `version/partition/"
+            "file`, expected `file/<version>/file`)."
```

Comment:
```suggestion
            "file`)."
```

Reply:
Switching this back, as the "expected" message is a confusing IMO if somebody is using a folder structure at the bottom level (e.g. as generated by Spark). The would think they already have version in `"path/to/folder/new/partition/version/partition/file"`, so the recommendation to have `file/<version>/file` seems ill-advised (really, they shouldn't be using a folder structure at the bottom, so let's not confuse them).